File: C-Plus-Plus/other/tower_of_hanoi.cpp
// C++ recursive function to
// solve tower of hanoi puzzle
#include <bits/stdc++.h>
using namespace std;

void towerOfHanoi(int n, char from_rod,
                  char to_rod, char aux_rod)
{
  //takes number of rod and rod labels as input
  // performs tower of hanoi algorithm
  // each update in position of disk is shown in output.
  if (n == 1)
    {
        cout << "Move disk 1 from rod " << from_rod <<
             " to rod " << to_rod << endl;
        return;
    }
    towerOfHanoi(n - 1, from_rod, aux_rod, to_rod);
    cout << "Move disk " << n << " from rod " << from_rod <<
         " to rod " << to_rod << endl;
    towerOfHanoi(n - 1, aux_rod, to_rod, from_rod);
}

// Driver code
int main()
{
    int n; // Number of disks
    cout << "Enter number of disks: ";
    cin >> n;
    
    towerOfHanoi(n, 'A', 'C', 'B'); // A, B and C are names of rods
    
    cout << "All disks have been moved to another rod!!!";
    return 0;
}

File: Java/other/TowerOfHanoi.java
import java.util.*;

class TowerOfHanoi {

    public static void shiftRings(int n, int fromRod, int toRod, int midRod) {
        if(n == 1) {
            System.out.println("Shifting disk 1 from rod " + fromRod + " to rod " + toRod);
            return;
        }
        shiftRings(n-1, fromRod, midRod, toRod);
        System.out.println("Shifting disk " + n + " from rod " + fromRod+ " to rod " + toRod);
        shiftRings(n-1, midRod, toRod, fromRod);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the number of rings:");
        int n = sc.nextInt();
        sc.close();

        shiftRings(n, 1, 3, 2);
    }
}
/**
 * Sample input/output: 
 * Enter the number of rings:
 * 4
 * Shifting disk 1 from rod 1 to rod 2
 * Shifting disk 2 from rod 1 to rod 3
 * Shifting disk 1 from rod 2 to rod 3
 * Shifting disk 3 from rod 1 to rod 2
 * Shifting disk 1 from rod 3 to rod 1
 * Shifting disk 2 from rod 3 to rod 2
 * Shifting disk 1 from rod 1 to rod 2
 * Shifting disk 4 from rod 1 to rod 3
 * Shifting disk 1 from rod 2 to rod 3
 * Shifting disk 2 from rod 2 to rod 1
 * Shifting disk 1 from rod 3 to rod 1
 * Shifting disk 3 from rod 2 to rod 3
 * Shifting disk 1 from rod 1 to rod 2
 * Shifting disk 2 from rod 1 to rod 3
 * Shifting disk 1 from rod 2 to rod 3
 * 
 * Time complexity: O(2^n)
 * Space complexity: O(n)
 */
